// @aws-sdk/client-sts/dist-es/commands/AssumeRoleWithWebIdentityCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "$Command", {
  enumerable: true,
  get: function () {
    return _smithyClient.Command;
  }
});
exports.AssumeRoleWithWebIdentityCommand = void 0;
var _middlewareEndpoint = require("@smithy/middleware-endpoint");
var _middlewareSerde = require("@smithy/middleware-serde");
var _smithyClient = require("@smithy/smithy-client");
var _types = require("@smithy/types");
var _models_ = require("../models/models_0");
var _Aws_query = require("../protocols/Aws_query");
class AssumeRoleWithWebIdentityCommand extends _smithyClient.Command {
  static getEndpointParameterInstructions() {
    return {
      UseGlobalEndpoint: {
        type: "builtInParams",
        name: "useGlobalEndpoint"
      },
      UseFIPS: {
        type: "builtInParams",
        name: "useFipsEndpoint"
      },
      Endpoint: {
        type: "builtInParams",
        name: "endpoint"
      },
      Region: {
        type: "builtInParams",
        name: "region"
      },
      UseDualStack: {
        type: "builtInParams",
        name: "useDualstackEndpoint"
      }
    };
  }
  constructor(input) {
    super();
    this.input = input;
  }
  resolveMiddleware(clientStack, configuration, options) {
    this.middlewareStack.use((0, _middlewareSerde.getSerdePlugin)(configuration, this.serialize, this.deserialize));
    this.middlewareStack.use((0, _middlewareEndpoint.getEndpointPlugin)(configuration, AssumeRoleWithWebIdentityCommand.getEndpointParameterInstructions()));
    const stack = clientStack.concat(this.middlewareStack);
    const {
      logger
    } = configuration;
    const clientName = "STSClient";
    const commandName = "AssumeRoleWithWebIdentityCommand";
    const handlerExecutionContext = {
      logger,
      clientName,
      commandName,
      inputFilterSensitiveLog: _models_.AssumeRoleWithWebIdentityRequestFilterSensitiveLog,
      outputFilterSensitiveLog: _models_.AssumeRoleWithWebIdentityResponseFilterSensitiveLog,
      [_types.SMITHY_CONTEXT_KEY]: {
        service: "AWSSecurityTokenServiceV20110615",
        operation: "AssumeRoleWithWebIdentity"
      }
    };
    const {
      requestHandler
    } = configuration;
    return stack.resolve(request => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
  }
  serialize(input, context) {
    return (0, _Aws_query.se_AssumeRoleWithWebIdentityCommand)(input, context);
  }
  deserialize(output, context) {
    return (0, _Aws_query.de_AssumeRoleWithWebIdentityCommand)(output, context);
  }
}
exports.AssumeRoleWithWebIdentityCommand = AssumeRoleWithWebIdentityCommand;
