// @aws-sdk/client-cognito-identity/dist-cjs/runtimeConfig.shared.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRuntimeConfig = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const url_parser_1 = require("@smithy/url-parser");
const util_base64_1 = require("@smithy/util-base64");
const util_utf8_1 = require("@smithy/util-utf8");
const endpointResolver_1 = require("./endpoint/endpointResolver");
const getRuntimeConfig = config => {
  return {
    apiVersion: "2014-06-30",
    base64Decoder: (config === null || config === void 0 ? void 0 : config.base64Decoder) ?? util_base64_1.fromBase64,
    base64Encoder: (config === null || config === void 0 ? void 0 : config.base64Encoder) ?? util_base64_1.toBase64,
    disableHostPrefix: (config === null || config === void 0 ? void 0 : config.disableHostPrefix) ?? false,
    endpointProvider: (config === null || config === void 0 ? void 0 : config.endpointProvider) ?? endpointResolver_1.defaultEndpointResolver,
    extensions: (config === null || config === void 0 ? void 0 : config.extensions) ?? [],
    logger: (config === null || config === void 0 ? void 0 : config.logger) ?? new smithy_client_1.NoOpLogger(),
    serviceId: (config === null || config === void 0 ? void 0 : config.serviceId) ?? "Cognito Identity",
    urlParser: (config === null || config === void 0 ? void 0 : config.urlParser) ?? url_parser_1.parseUrl,
    utf8Decoder: (config === null || config === void 0 ? void 0 : config.utf8Decoder) ?? util_utf8_1.fromUtf8,
    utf8Encoder: (config === null || config === void 0 ? void 0 : config.utf8Encoder) ?? util_utf8_1.toUtf8
  };
};
exports.getRuntimeConfig = getRuntimeConfig;
