// @aws-sdk/client-sts/dist-es/runtimeConfig.shared.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRuntimeConfig = void 0;
var _smithyClient = require("@smithy/smithy-client");
var _urlParser = require("@smithy/url-parser");
var _utilBase = require("@smithy/util-base64");
var _utilUtf = require("@smithy/util-utf8");
var _endpointResolver = require("./endpoint/endpointResolver");
const getRuntimeConfig = config => {
  return {
    apiVersion: "2011-06-15",
    base64Decoder: (config === null || config === void 0 ? void 0 : config.base64Decoder) ?? _utilBase.fromBase64,
    base64Encoder: (config === null || config === void 0 ? void 0 : config.base64Encoder) ?? _utilBase.toBase64,
    disableHostPrefix: (config === null || config === void 0 ? void 0 : config.disableHostPrefix) ?? false,
    endpointProvider: (config === null || config === void 0 ? void 0 : config.endpointProvider) ?? _endpointResolver.defaultEndpointResolver,
    extensions: (config === null || config === void 0 ? void 0 : config.extensions) ?? [],
    logger: (config === null || config === void 0 ? void 0 : config.logger) ?? new _smithyClient.NoOpLogger(),
    serviceId: (config === null || config === void 0 ? void 0 : config.serviceId) ?? "STS",
    urlParser: (config === null || config === void 0 ? void 0 : config.urlParser) ?? _urlParser.parseUrl,
    utf8Decoder: (config === null || config === void 0 ? void 0 : config.utf8Decoder) ?? _utilUtf.fromUtf8,
    utf8Encoder: (config === null || config === void 0 ? void 0 : config.utf8Encoder) ?? _utilUtf.toUtf8
  };
};
exports.getRuntimeConfig = getRuntimeConfig;
