// @aws-sdk/client-sts/dist-es/STS.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STS = void 0;
var _smithyClient = require("@smithy/smithy-client");
var _AssumeRoleCommand = require("./commands/AssumeRoleCommand");
var _AssumeRoleWithSAMLCommand = require("./commands/AssumeRoleWithSAMLCommand");
var _AssumeRoleWithWebIdentityCommand = require("./commands/AssumeRoleWithWebIdentityCommand");
var _DecodeAuthorizationMessageCommand = require("./commands/DecodeAuthorizationMessageCommand");
var _GetAccessKeyInfoCommand = require("./commands/GetAccessKeyInfoCommand");
var _GetCallerIdentityCommand = require("./commands/GetCallerIdentityCommand");
var _GetFederationTokenCommand = require("./commands/GetFederationTokenCommand");
var _GetSessionTokenCommand = require("./commands/GetSessionTokenCommand");
var _STSClient = require("./STSClient");
const commands = {
  AssumeRoleCommand: _AssumeRoleCommand.AssumeRoleCommand,
  AssumeRoleWithSAMLCommand: _AssumeRoleWithSAMLCommand.AssumeRoleWithSAMLCommand,
  AssumeRoleWithWebIdentityCommand: _AssumeRoleWithWebIdentityCommand.AssumeRoleWithWebIdentityCommand,
  DecodeAuthorizationMessageCommand: _DecodeAuthorizationMessageCommand.DecodeAuthorizationMessageCommand,
  GetAccessKeyInfoCommand: _GetAccessKeyInfoCommand.GetAccessKeyInfoCommand,
  GetCallerIdentityCommand: _GetCallerIdentityCommand.GetCallerIdentityCommand,
  GetFederationTokenCommand: _GetFederationTokenCommand.GetFederationTokenCommand,
  GetSessionTokenCommand: _GetSessionTokenCommand.GetSessionTokenCommand
};
class STS extends _STSClient.STSClient {}
exports.STS = STS;
(0, _smithyClient.createAggregatedClient)(commands, STS);
