// @aws-sdk/client-sso/dist-es/commands/GetRoleCredentialsCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "$Command", {
  enumerable: true,
  get: function () {
    return _smithyClient.Command;
  }
});
exports.GetRoleCredentialsCommand = void 0;
var _middlewareEndpoint = require("@smithy/middleware-endpoint");
var _middlewareSerde = require("@smithy/middleware-serde");
var _smithyClient = require("@smithy/smithy-client");
var _types = require("@smithy/types");
var _models_ = require("../models/models_0");
var _Aws_restJson = require("../protocols/Aws_restJson1");
class GetRoleCredentialsCommand extends _smithyClient.Command {
  static getEndpointParameterInstructions() {
    return {
      UseFIPS: {
        type: "builtInParams",
        name: "useFipsEndpoint"
      },
      Endpoint: {
        type: "builtInParams",
        name: "endpoint"
      },
      Region: {
        type: "builtInParams",
        name: "region"
      },
      UseDualStack: {
        type: "builtInParams",
        name: "useDualstackEndpoint"
      }
    };
  }
  constructor(input) {
    super();
    this.input = input;
  }
  resolveMiddleware(clientStack, configuration, options) {
    this.middlewareStack.use((0, _middlewareSerde.getSerdePlugin)(configuration, this.serialize, this.deserialize));
    this.middlewareStack.use((0, _middlewareEndpoint.getEndpointPlugin)(configuration, GetRoleCredentialsCommand.getEndpointParameterInstructions()));
    const stack = clientStack.concat(this.middlewareStack);
    const {
      logger
    } = configuration;
    const clientName = "SSOClient";
    const commandName = "GetRoleCredentialsCommand";
    const handlerExecutionContext = {
      logger,
      clientName,
      commandName,
      inputFilterSensitiveLog: _models_.GetRoleCredentialsRequestFilterSensitiveLog,
      outputFilterSensitiveLog: _models_.GetRoleCredentialsResponseFilterSensitiveLog,
      [_types.SMITHY_CONTEXT_KEY]: {
        service: "SWBPortalService",
        operation: "GetRoleCredentials"
      }
    };
    const {
      requestHandler
    } = configuration;
    return stack.resolve(request => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
  }
  serialize(input, context) {
    return (0, _Aws_restJson.se_GetRoleCredentialsCommand)(input, context);
  }
  deserialize(output, context) {
    return (0, _Aws_restJson.de_GetRoleCredentialsCommand)(output, context);
  }
}
exports.GetRoleCredentialsCommand = GetRoleCredentialsCommand;
